runtime/common/src/assigned_slots.rs

```diff
@@ -0,0 +1,1086 @@
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: configuration::Config + paras::Config + slots::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Origin for assigning slots
+		type AssignSlotOrigin: EnsureOrigin<<Self as frame_system::Config>::Origin>;
+
+		/// The type representing the leasing system.
+		type Leaser: Leaser<AccountId = Self::AccountId, LeasePeriod = Self::BlockNumber>;
+
+		/// The number of lease periods a permanent parachain slot lasts
+		#[pallet::constant]
+		type PermanentSlotLeasePeriodLength: Get<u32>;
+
+		/// The number of lease periods a temporary parachain slot lasts
+		#[pallet::constant]
+		type TemporarySlotLeasePeriodLength: Get<u32>;
+
+		/// The max number of permanent slots that can be assigned
+		#[pallet::constant]
+		type MaxPermanentSlots: Get<u32>;
+
+		/// The max number of temporary slots to be scheduled per lease periods
+		#[pallet::constant]
+		type MaxTemporarySlotPerLeasePeriod: Get<u32>;
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn permanent_slots)]
+	pub type PermanentSlots<T: Config> =
+		StorageMap<_, Twox64Concat, ParaId, Option<LeasePeriodOf<T>>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn permanent_slot_count)]
+	pub type PermanentSlotCount<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn temporary_slots)]
+	pub type TemporarySlots<T: Config> = StorageMap<
+		_,
+		Twox64Concat,
+		ParaId,
+		Option<ParachainTemporarySlot<T::AccountId, LeasePeriodOf<T>>>,
+		ValueQuery,
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A para was assigned a permanent parachain slot
+		PermanentSlotAssigned(ParaId),
+		/// A para was assigned a temporary parachain slot
+		TemporarySlotAssigned(ParaId),
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The specified parachain or parathread is not registered.
+		ParaDoesntExist,
+		/// Not a parathread.
+		NotParathread,
+		/// Cannot upgrade parathread.
+		CannotUpgrade,
+		/// Cannot downgrade parachain.
+		CannotDowngrade,
+		/// Permanent or Temporary slot already assigned.
+		SlotAlreadyAssigned,
+		/// Permanent or Temporary slot has not been assigned.
+		SlotNotAssigned,
+		/// An ongoing lease already exists.
+		OngoingLeaseExists,
+		// Maximum number of permanent slots exceeded
+		MaxPermanentSlotsExceeded,
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_initialize(n: T::BlockNumber) -> Weight {
+			let lease_period = Self::lease_period();
+			let lease_period_index = Self::lease_period_index();
+			if (n % lease_period).is_zero() {
+				Self::manage_lease_period_start(lease_period_index)
+			} else {
+				0
+			}
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Assign a permanent parachain slot
+		#[pallet::weight((1_000, DispatchClass::Operational))]
```

Comment:
this is not benchmarked because it will not end up in production?

Reply:
Added weight of the delegated call on the `slots` pallet e.g. `1_000 + <T as slots::Config>::WeightInfo::force_lease()`. I think it's a reasonable approximation.